content/en/blog/_posts/2022-04-19-non-graceful-node-shutdown.md

```diff
@@ -0,0 +1,63 @@
+---
+
+**Authors** Xing Yang and Yassine Tijani (VMware)
+
+Kubernetes v1.24 introduces alpha support for [Non-Graceful Node Shutdown](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2268-non-graceful-shutdown). This feature allows stateful workloads to failover to a different node after the original node is shutdown or in a non-recoverable state such as the hardware failure or broken OS.
+
+## How is this different from Graceful Node Shutdown
+
+You might have heard about the [Graceful Node Shutdown](https://kubernetes.io/blog/2021/04/21/graceful-node-shutdown-beta/) feature and are wondering how the Non-Graceful Node Shutdown feature is different from that. Graceful Node Shutdown introduced a way to detect a node shutdown and handle it gracefully. A Node Shutdown can be `Graceful` only if the node shutdown action can be detected by Kubelet's Node Shutdown Manager. However, there are cases where a node shutdown action may not be detected by Kubelet's Node Shutdown Manager. This could happen either because the shutdown command does not trigger the inhibitor locks mechanism used by Kubelet or because of a user error, i.e., the `ShutdownGracePeriod` and `ShutdownGracePeriodCriticalPods` are not configured properly.
+
+When a node is shutdown but not detected by Kubelet's Node Shutdown Manager, the pods that are part of a StatefulSet will be stuck in terminating status on the shutdown node and cannot move to a new running node. As a result, the application running on the StatefulSet cannot function properly. If the original shutdown node comes up, the pods will be deleted by Kubelet and new pods will be created on a different running node. If the original shutdown node does not come up, these pods will be stuck in terminating status on the shutdown node forever.
+
+```
+# kubectl get pod -o wide
+NAME    READY   STATUS        RESTARTS   AGE   IP           NODE                      NOMINATED NODE   READINESS GATES
+web-0   1/1     Running       0          100m   10.244.2.4   k8s-node-876-1639279816   <none>           <none>
+web-1   1/1     Terminating   0          100m   10.244.1.3   k8s-node-433-1639279804   <none>           <none>
+```
+
+## How does it work
+
+To use the Non-Graceful Node Shutdown feature, the `NodeOutOfServiceVolumeDetach` feature gate needs to be enabled on `kube-controller-manager`.
+
+In the case of a node shutdown, if the graceful shutdown is not working or the node is in a non-recoverable state due to a hardware failure or broken OS, the user can manually add a `out-of-service’ taint on the Node. For example, this can be `node.kubernetes.io/out-of-service=nodeshutdown:NoExecute` or `node.kubernetes.io/out-of-service=nodeshutdown:NoSchedule`. This will trigger pods on the node to be forcefully deleted if there are no matching tolerations on the pods. Persistent volumes attached to the shutdown node will be detached, and new pods will be created successfully on a different running node.
```

Comment:
```suggestion
In the case of a node shutdown, you can manually taint that node as out of service. You should make certain that
the node is truly shutdown (and won't restart during maintenance) before you add that taint. You could add that
taint following a clean shutdown that the kubelet did not detect and handle in advance; another case where you
can use that taint is when a node has failed and you are taking it out of service. 
The values you set for that taint can be `node.kubernetes.io/out-of-service=nodeshutdown: "NoExecute"`
or `node.kubernetes.io/out-of-service=nodeshutdown:" NoSchedule"`.
provided you have enabled the feature gate as I mentioned earlier, setting the out-of-service taint on a Node
means that pods on the node will be deleted (unless you happened to have set matching tolerations on the pods).
Persistent volumes attached to the shutdown node will be detached, and for StatefulSets, replacement pods will
be created successfully on a different running node.
```

Reply:
I prefer original sentence: 
```
In the case of a node shutdown, if the graceful shutdown is not working or the node is in a non-recoverable state due to a hardware failure or broken OS
```
because even if when the node is not truly shutdown, when the node has hardware failure or its OS is broken, pods on the node will be stuck in terminating status.

FYI: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2268-non-graceful-shutdown#proposal